### PR TITLE
drivers: GPIO Added GPIO for SAM family.

### PR DESCRIPTION
**Pull request by fallrisk**
Added GPIO driver for SAM family of SoCs.

Signed-off-by: Justin Watson <jwatson5@gmail.com>

### CODE REVIEW

**Comment by fallrisk** [comment]
I tested these changes on the SAM E70 Xplained hardware and the SAM4S Xplained Hardware. I used the samples "blinky" and "button".

**Comment by fallrisk** [comment]
@galak Can you get @mnkp on the reviewer list?

**Comment by galak** [comment]
@fallrisk I foobar'd this PR.  I was trying to rebase it and accidentally pushed the wrong branch.  Can you push the GPIO change again.  Sorry about that

**Review by dbkinder — DISMISSED** [review]

**Review by galak — COMMENTED** [review]
In the dts.fixup you go to PORTE in the driver looks like only up to PORTC.

**Comment by fallrisk** [comment]
SAM E70 goes up to E.

**Comment by galak** [comment]
So shouldn't the driver go up to E as well?

**Comment by fallrisk** [comment]
Yes, will fix that in a few mins.

**Review by galak — CHANGES_REQUESTED** [review]
peripheral-id should be coming from device tree.

File: drivers/gpio/gpio_sam.c [hunk 1]
```diff
@@ -0,0 +1,390 @@
...
+	return 0;
+}
+
+/* PORT A */
+#ifdef CONFIG_GPIO_SAM_PORTA_BASE_ADDRESS
+static void port_a_sam_config_func(struct device *dev);
+
+static const struct gpio_sam_config port_a_sam_config = {
+	.regs = (Pio *)CONFIG_GPIO_SAM_PORTA_BASE_ADDRESS,
+	.periph_id = ID_PIOA,
```
Comment (galak) [inline drivers/gpio/gpio_sam.c]: should come from DT

File: dts/arm/atmel/sam4s.dtsi [hunk 2]
```diff
@@ -87,6 +87,27 @@
 			status = "disabled";
 			label = "USART_1";
 		};
+
+		porta: gpio@400E0E00 {
+			compatible = "atmel,sam-gpio";
+			reg = <0x400E0E00 0x190>;
+			interrupts = <11 1>;
+			label = "PORTA";
```
Comment (galak) [inline dts/arm/atmel/sam4s.dtsi]: need peripheral-id = <>

File: dts/bindings/gpio/atmel.sam-gpio.yaml [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+      type: compound
+      category: required
+      description: required interrupts
+      generation: define
+
+    label:
+      type: string
+      category: required
+      description: Human readable string describing the device (used by Zephyr for API name)
+      generation: define
```
Comment (galak) [inline dts/bindings/gpio/atmel.sam-gpio.yaml]: add peripheral-id

**Comment by galak** [comment]
Looks like you are out of date w/upstream arm branch.  I rebased and pushed an update against it.

File: drivers/gpio/gpio_sam.c [hunk 4]
```diff
@@ -0,0 +1,390 @@
...
+	}
+
+	/* Setup Pull-down resistor. */
+	if ((flags & GPIO_PUD_MASK) == GPIO_PUD_PULL_DOWN) {
+		pio->PIO_PPDER = mask;
+	} else {
+		pio->PIO_PPDDR = mask;
+	}
+
+#ifdef SOC_SERIES_SAM3X
```
Comment (galak) [inline drivers/gpio/gpio_sam.c]: nit: but since you have an #elif, this usually from a style point of view is #if defined(SOC_SERIES_SAM3X)

**Review by galak — APPROVED** [review]

**Pull request merged**